usage/configuration.md
# Configuration

The only **required** configuration option is `baseUrl` which will be used for API calls to your Laravel API, so you can start using the module with the following definition:

{% code title="nuxt.config.ts" fullWidth="false" %}
```typescript
export default defineNuxtConfig({
    modules: ['nuxt-auth-sanctum'],

    sanctum: {
        baseUrl: 'http://localhost:80', // Laravel API
    },
});
```
{% endcode %}

### Available options

For any additional configurations, you can adjust the next list of available parameters:

| Parameter                            | Description                                                                                                                          | Default value            |
| ------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------ | ------------------------ |
| baseUrl                              | The base URL of the Laravel API                                                                                                      | undefined                |
| mode                                 | Authentication mode to work with Laravel API. Supported values - `cookie`, `token`.                                                  | cookie                   |
| origin                               | The URL of the current application to use in Referrer header                                                                         | `useRequestUrl().origin` |
| userStateKey                         | The key to use to store the user identity in the `useState` variable.                                                                | sanctum.user.identity    |
| redirectIfAuthenticated              | Determine whether to redirect the user if it is already authenticated on a login attempt.                                            | false                    |
| redirectIfUnauthenticated            | Determine whether to redirect when the user got unauthenticated on any API request.                                                  | false                    |
| endpoints.csrf                       | The endpoint to request a new CSRF token                                                                                             | `/sanctum/csrf-cookie`   |
| endpoints.login                      | The endpoint to send user credentials to authenticate                                                                                | `/login`                 |
| endpoints.logout                     | The endpoint to destroy current user session                                                                                         | `/logout`                |
| endpoints.user                       | The endpoint to fetch current user data                                                                                              | `/api/user`              |
| csrf.cookie                          | Name of the CSRF cookie to extract from server response                                                                              | XSRF-TOKEN               |
| csrf.header                          | Name of the CSRF header to pass from client to server                                                                                | X-XSRF-TOKEN             |
| client.retry                         | The number of times to retry a request when it fails                                                                                 | false                    |
| client.initialRequest                | Determines whether to request the user identity on plugin initialization                                                             | true                     |
| redirect.keepRequestedRoute          | Determines whether to keep the requested route when redirecting after login                                                          | false                    |
| redirect.onLogin                     | Route to redirect to when user is authenticated. If set to false, do nothing                                                         | `/`                      |
| redirect.onLogout                    | Route to redirect to when user is not authenticated. If set to false, do nothing                                                     | `/`                      |
| redirect.onAuthOnly                  | Route to redirect to when user has to be authenticated. If set to false, do nothing                                                  | `/login`                 |
| redirect.onGuestOnly                 | Route to redirect to when user has to be a guest. If set to false, do nothing                                                        | `/`                      |
| globalMiddleware.enabled             | Determines whether the global middleware is enabled                                                                                  | false                    |
| globalMiddleware.prepend             | Determines whether to allow 404 pages without authentication                                                                         | false                    |
| globalMiddleware.allow404WithoutAuth | Determines whether to allow 404 page without authentication                                                                          | true                     |
| logLevel                             | The level to use for the logger. More details [here](../advanced/logging.md).                                                        | 3                        |
| appendPlugin                         | Determines whether to append the plugin to the Nuxt application. More details [here](https://nuxt.com/docs/api/kit/plugins#options). | false                    |

For more details, please check the source code - [options.ts](https://github.com/manchenkoff/nuxt-auth-sanctum/blob/main/src/runtime/types/options.ts).

### Overrides

You can override any of these options in the `nuxt.config.ts` file:

```ts
export default defineNuxtConfig({
    modules: ['nuxt-auth-sanctum'],

    sanctum: {
        baseUrl: 'http://localhost:80', // Your Laravel API
        redirect: {
            onLogin: '/dashboard', // Custom route after successful login
        },
    },
});
```

### RuntimeConfig

Module configuration is exposed to `runtimeConfig` property of your Nuxt app, so you can override either in `sanctum` module config or `runtimeConfig.public.sanctum` property.

```ts
export default defineNuxtConfig({
    modules: ['nuxt-auth-sanctum'],

    runtimeConfig: {
        public: {
            sanctum: {
                baseUrl: 'http://localhost:80',
            },
        },
    },
});
```

### Environment variables

It is possible to override options via environment variables too. It might be useful when you want to use `.env` file to provide `baseUrl` for Laravel API.

And here is what it will look like in `.env` file:

`NUXT_PUBLIC_SANCTUM_BASE_URL='http://localhost:80'`

### Configuration example

Here is an example of a full module configuration

```typescript
sanctum: {
    mode: 'cookie',
    userStateKey: 'sanctum.user.identity',
    redirectIfAuthenticated: false,
    redirectIfUnauthenticated: false,
    endpoints: {
        csrf: '/sanctum/csrf-cookie',
        login: '/login',
        logout: '/logout',
        user: '/api/user',
    },
    csrf: {
        cookie: 'XSRF-TOKEN',
        header: 'X-XSRF-TOKEN',
    },
    client: {
        retry: false,
        initialRequest: true,
    },
    redirect: {
        keepRequestedRoute: false,
        onLogin: '/',
        onLogout: '/',
        onAuthOnly: '/login',
        onGuestOnly: '/',
    },
    globalMiddleware: {
        enabled: false,
        allow404WithoutAuth: true,
    },
    logLevel: 3,
    appendPlugin: false,
}
```


composables/usesanctumconfig.md
# useSanctumConfig

This composable provides quick access to the module configuration instead of using `useRuntimeConfig` and several keys like `public.sanctum`.&#x20;

Take a look at the following example

{% code fullWidth="false" %}
```ts
const config = useSanctumConfig();

console.log(config.baseUrl); // runtimeConfig.public.sanctum.baseUrl
```
{% endcode %}

More details about the configuration structure can be found [here](../usage/configuration.md).


CHANGELOG.md
# Changelog


## v0.5.8

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.7...v0.5.8)

### 🩹 Fixes

- Use unique key for sanctum fetch across csr/ssr ([0d43ea6](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/0d43ea6))

### ❤️ Contributors

- Manchenkoff ([@manchenkoff](http://github.com/manchenkoff))

## v0.5.7

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.6...v0.5.7)

### 🩹 Fixes

- Migrated from useFetch to useAsyncData ([79cb27d](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/79cb27d))

### 🏡 Chore

- Dependencies upgrade ([2bbe90c](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/2bbe90c))
- **dev-deps:** Upgraded dev dependencies ([6a629df](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/6a629df))
- **dev-deps:** Upgraded nuxt devtools ([bc3b601](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/bc3b601))
- **security:** Upgraded esbuild to avoid security issues ([82ad135](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/82ad135))

### ❤️ Contributors

- Manchenkoff ([@manchenkoff](http://github.com/manchenkoff))

## v0.5.6

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.5...v0.5.6)

### 🚀 Enhancements

- Added new composables to make requests ([a6bfc2f](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/a6bfc2f))

### 🩹 Fixes

- Added type hints for fetch composables ([bbb9e04](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/bbb9e04))

### ❤️ Contributors

- Manchenkoff ([@manchenkoff](http://github.com/manchenkoff))

## v0.5.5

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.4...v0.5.5)

### 🩹 Fixes

- Use npm keys in actions ([9bc984f](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/9bc984f))
- Deduplicate `set-cookie` headers on SSR response ([9d957e2](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/9d957e2))
- Downgraded typescript to fix module-builder ([f28daf8](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/f28daf8))

### 🏡 Chore

- **deps-dev:** Bump @nuxt/devtools from 1.6.4 to 1.7.0 ([b58ebc4](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/b58ebc4))
- **deps-dev:** Bump @nuxt/eslint-config from 0.7.3 to 0.7.4 ([0f79c42](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/0f79c42))
- **deps-dev:** Bump @typescript-eslint/eslint-plugin ([0cae715](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/0cae715))
- **deps-dev:** Bump @nuxt/schema from 3.14.1592 to 3.15.0 ([77c79f6](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/77c79f6))
- **deps-dev:** Bump @nuxt/test-utils from 3.15.1 to 3.15.4 ([2cc446b](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/2cc446b))
- **deps-dev:** Bump vitest in the npm_and_yarn group ([b64a354](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/b64a354))
- **deps-dev:** Bump @nuxt/schema from 3.15.2 to 3.15.4 ([c2dd453](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/c2dd453))
- **deps-dev:** Bump @nuxt/eslint-config from 0.7.4 to 1.0.0 ([a435516](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/a435516))
- Upgraded nuxt packages ([e0c5bb9](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/e0c5bb9))

### ❤️ Contributors

- Manchenkoff ([@manchenkoff](http://github.com/manchenkoff))

## v0.5.4

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.3...v0.5.4)

### 🩹 Fixes

- Use relative paths for injected types ([63696ff](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/63696ff))

### ❤️ Contributors

- Denis Mamaev ([@ExileofAranei](http://github.com/ExileofAranei))

## v0.5.3

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.2...v0.5.3)

### 🩹 Fixes

- Expose plugin name to be used in dependsOn ([c2d930b](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/c2d930b))

### 🏡 Chore

- **deps-dev:** Bump @nuxt/schema from 3.14.159 to 3.14.1592 ([83be062](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/83be062))
- **deps-dev:** Bump nuxt from 3.14.159 to 3.14.1592 ([0469f64](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/0469f64))
- **deps-dev:** Bump @typescript-eslint/eslint-plugin ([f619be8](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/f619be8))
- **deps-dev:** Bump eslint from 9.15.0 to 9.16.0 ([464c8ed](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/464c8ed))
- **deps-dev:** Bump @nuxt/eslint-config from 0.7.0 to 0.7.2 ([38c1c46](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/38c1c46))
- **deps-dev:** Bump @nuxt/devtools from 1.6.1 to 1.6.3 ([31a8027](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/31a8027))
- **deps-dev:** Bump @typescript-eslint/eslint-plugin ([1d47a17](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/1d47a17))
- **deps-dev:** Bump vitest from 2.1.5 to 2.1.8 ([63fa544](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/63fa544))
- **deps-dev:** Bump @nuxt/test-utils from 3.14.4 to 3.15.1 ([7c21b35](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/7c21b35))
- **deps-dev:** Bump @types/node from 22.10.1 to 22.10.2 ([b1d429b](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/b1d429b))
- **deps-dev:** Bump @typescript-eslint/eslint-plugin ([3067b61](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/3067b61))
- **deps-dev:** Bump eslint from 9.16.0 to 9.17.0 ([ce17914](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/ce17914))
- **deps-dev:** Bump @nuxt/eslint-config from 0.7.2 to 0.7.3 ([2ff5e99](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/2ff5e99))
- Upgraded playground to nuxt v4 structure ([056b488](https://github.com/manchenkoff/nuxt-auth-sanctum/commit/056b488))

### ❤️ Contributors

- Manchenkoff ([@manchenkoff](http://github.com/manchenkoff))

## v0.5.2

[compare changes](https://github.com/manchenkoff/nuxt-auth-sanctum/compare/v0.5.1...v0.5.2)

### 🩹 Fixes


src/runtime/types/options.ts
/**
 * Definition of Laravel Sanctum endpoints to be used by the client.
 */
export interface SanctumEndpoints {
  /**
   * The endpoint to request a new CSRF token.
   * @default '/sanctum/csrf-cookie'
   */
  csrf: string
  /**
   * The endpoint to send user credentials to authenticate.
   * @default '/login'
   */
  login: string
  /**
   * The endpoint to destroy current user session.
   * @default '/logout'
   */
  logout: string
  /**
   * The endpoint to fetch current user data.
   * @default '/api/user'
   */
  user: string
}

/**
 * CSRF token specific options.
 */
export interface CsrfOptions {
  /**
   * Name of the CSRF cookie to extract from server response.
   * @default 'XSRF-TOKEN'
   */
  cookie: string
  /**
   * Name of the CSRF header to pass from client to server.
   * @default 'X-XSRF-TOKEN'
   */
  header: string
}

/**
 * OFetch client specific options.
 */
export interface ClientOptions {
  /**
   * The number of times to retry a request when it fails.
   * @default false
   */
  retry: number | boolean
  /**
   * Determines whether to request the user identity on plugin initialization.
   * @default true
   */
  initialRequest: boolean
}

/**
 * Behavior of the plugin redirects when user is authenticated or not.
 */
export interface RedirectOptions {
  /**
   * Determines whether to keep the requested route when redirecting after login.
   * @default false
   */
  keepRequestedRoute: boolean
  /**
   * Route to redirect to when user is authenticated.
   * If set to false, do nothing.
   * @default '/'
   */
  onLogin: string | false
  /**
   * Route to redirect to when user is not authenticated.
   * If set to false, do nothing.
   * @default '/'
   */
  onLogout: string | false
  /**
   * Route to redirect to when user has to be authenticated.
   * If set to false, the plugin will throw an 403 error.
   * @default '/login'
   */
  onAuthOnly: string | false
  /**
   * Route to redirect to when user has to be a guest.
   * If set to false, the plugin will throw an 403 error.
   * @default '/'
   */
  onGuestOnly: string | false
}

/**
 * Configuration of the global application-wide middleware.
 */
export interface GlobalMiddlewareOptions {
  /**
   * Determines whether the global middleware is enabled.
   * @default false
   */
  enabled: boolean
  /**
   * Determines whether the global middleware is prepended.
   * @default false
   */
  prepend: boolean
  /**
   * Determines whether to allow 404 pages without authentication.
   * @default true
   */
  allow404WithoutAuth: boolean
}

/**
 * Options to be passed to the plugin.
 */
export interface ModuleOptions {
  /**
   * The base URL of the Laravel API.
   * @default 'http://localhost:80'
   */
  baseUrl: string
  /**
   * The mode to use for authentication.
   * @default 'cookie'
   */
  mode: 'cookie' | 'token'
  /**
   * The URL of the current application to use in Referrer header. (Optional)
   * @default useRequestURL().origin
   */
  origin?: string
  /**
   * The key to use to store the user identity in the `useState` variable.
   * @default 'sanctum.user.identity'
   */
  userStateKey: string
  /**
   * Determine whether to redirect the user if it is already authenticated on a login attempt.
   * @default false
   */
  redirectIfAuthenticated: boolean
  /**
   * Determine whether to redirect when the user got unauthenticated on any API request.
   * @default false
   */
  redirectIfUnauthenticated: boolean
  /**
   * Laravel Sanctum endpoints to be used by the client.
   */
  endpoints: Partial<SanctumEndpoints>
  /**
   * CSRF token specific options.
   */
  csrf: Partial<CsrfOptions>
  /**
   * OFetch client specific options.
   */
  client: Partial<ClientOptions>
  /**
   * Behavior of the plugin redirects when user is authenticated or not.
   */
  redirect: Partial<RedirectOptions>
  /**
   * Behavior of the global middleware.
   */
  globalMiddleware: Partial<GlobalMiddlewareOptions>
  /**
   * The log level to use for the logger.
   *
   * 0: Fatal and Error
   * 1: Warnings
   * 2: Normal logs
   * 3: Informational logs
   * 4: Debug logs
   * 5: Trace logs
   *
   * More details at https://github.com/unjs/consola?tab=readme-ov-file#log-level
   * @default 3
   */
  logLevel: number
  /**
   * Determines whether to append the plugin to the Nuxt application.
   * Be default, Nuxt prepends the plugin to load it before the application modules.
   * @default false
   * @see https://nuxt.com/docs/api/kit/plugins#options
   */
  appendPlugin: boolean
}
